File: 2020/7.pl
#!perl -lp0
sub x{my$t=/^@_ (.*)/m;for$c($1=~/\d+ \w+ \w+/g){$c=~$";$t+=$c*x($')}$t};
$\.=x($x="shiny gold")-1;$a++while s/^(\w+ \w+) .*?($x)/$x.="|$1"/em;$_=$a

File: 2020/7.js
#!/usr/bin/env node
const A = require('../advent');
const l = console.log
const content = require('fs').readFileSync('7.input.txt','utf8');

const input = A.parse(content, /(.+)/g);
function solve(input, part2) {
  const g = A.graph();
  for (const line of input) {
    const [a] = A.parse(line, /(\w+ \w+)/);
    const b = A.parse(line, /(\d+) (\w+ \w+)/g);
    for (const v of b) g.connect(a, v[1], v[0]);
  }
  return (part2 ?
    g.dfsReduce('shiny gold', n => 1 + A.sum(n.map(([e, v]) => e * v))) :
    g.bfsRev('shiny gold').size) - 1;
}
l(solve(input));
l(solve(input, true));
